#!/bin/bash

# Christin M. Hong
# Last modified: 2015-10
# Lab of Soumya Raychaudhuri, Harvard Medical School

# Quantifying allele-specific expression after calling variants with GATK Haplotype Caller and mapping reads with Subread.


#####

# VCF file with filtered variants with IDs
varsFilt=ASE02f10_${pathInd}_hets_allFilts.ids.vcf

# Input and output identifiers
regex="ASE02f13_3subreadAligned_(SR.*).bam"
[[ ${1} =~ ${regex} ]]

out1Pre=ASE02f14_1SamSort_${pathInd}_"${BASH_REMATCH[1]}"
out1=ASE02f14_1SamSort_${pathInd}_"${BASH_REMATCH[1]}".sorted.bam
out2=ASE02f14_2SamPileup_${pathInd}_"${BASH_REMATCH[1]}".pileup.txt
out3=ASE02f14_3SamParse_${pathInd}_"${BASH_REMATCH[1]}".parsed_pileup.txt
out4=ASE02f14_4SamSNP_${pathInd}_"${BASH_REMATCH[1]}".SNP.txt
out5=ASE02f14_5SamASE_${pathInd}_"${BASH_REMATCH[1]}".ASE.txt


#####

# Sort bam files generated by mapping reads with subread to the personalized masked genome.
samtools sort \
     -o ${out1} \
     -O bam \
     -T ${out1Pre} \
     ${1} 


# Index aligned libraries
samtools index ${out1}


# Quantify mapped reads
samtools mpileup \
     --fasta-ref ${fileRef} \
     --positions ${varsFilt} \
     --output-MQ \
     --no-BAQ \
     ${out1} \
     > ${out2}     


# Quantify mapped reads over filtered variant sites from Haplotype Caller
perl ${scriptASE} -parse_pileup \
     -sp ${out2} \
     -vcf ${varsFilt} \
     > ${out3}


# Create file with gene information for SNP locations
perl ${scriptASE} --get_gene_snp_mapping_file \
     --use_gencode \
     -is ${varsFilt} \
     > ${out4}
     

# Add gene information (gene SNP mapping file) to read quantification (parsed pileup), format more nicely, and perform binomial test for ASE. 
perl ${scriptASE} --calculate_ase_normal \
     -pp ${out3} \
     -is ${varsFilt} \
     -gf ${out4} \
     -mrs 10 \
     -mq 20 \
     -bq 10 \
     > ${out5}


# Notes
     # perl ${scriptASE} --calculate_ase_normal: We don't use the stats from their binomial test, but the formatting's helpful for downstream analysis.
          # -r 0.5: Reference ratio.  Removed this option because it wasn't being recognized by the perl script and the default ratio is already 0.5.
          # -mrs 10: Min reads/site
          # -mq 20: Min mapping quality score
          # -bq 10: Min base quality

